IPNS.md

```diff
@@ -29,95 +36,258 @@ Each time a file is modified, its content address changes. As a consequence, the
 
 IPNS is based on [SFS](http://en.wikipedia.org/wiki/Self-certifying_File_System). It consists of a PKI namespace, where a name is simply the hash of a public key. As a result, whoever controls the private key has full control over the name. Accordingly, records are signed by the private key and then distributed across the network (in IPFS, via the routing system). This is an egalitarian way to assign mutable names on the Internet at large, without any centralization whatsoever, or certificate authorities.
 
-## IPNS Record
+## IPNS Keys
 
-An IPNS record is a data structure containing the following fields:
-
-- 1. **Value** (bytes)
-  - It can be any path, such as a path to another IPNS record, a `dnslink` path (eg. `/ipns/example.com`) or an IPFS path (eg. `/ipfs/Qm...`)
-- 2. **Validity** (bytes)
-  - Expiration date of the record using [RFC3339](https://www.ietf.org/rfc/rfc3339.txt) with nanoseconds precision.
-  - Note: Currently, the expiration date is the only available type of validity.
-- 3. **Validity Type** (uint64)
-   - Allows us to define the conditions under which the record is valid.
-   - Only supports expiration date with `validityType = 0` for now.
-- 4. **Signature** (bytes)
-  - Concatenate value, validity field and validity type
-  - Sign the concatenation result with the provided private key
-  - Note: Once we add new validity types, the signature must be changed. More information on [ipfs/notes#249](https://github.com/ipfs/notes/issues/249)
-- 5. **Sequence** (uint64)
-  - Represents the current version of the record (starts at 0)
-- 6. **Public Key** (bytes)
-  - Public key used to sign this record
-  - Note: The public key **must** be included if it cannot be extracted from the peer ID (reference [libp2p/specs#100](https://github.com/libp2p/specs/pull/100/files)).
-- 7. **ttl** (uint64)
-  - A hint for how long the record should be cached before going back to, for instance the DHT, in order to check if it has been updated.
+### Key Types
+
+Implementations MUST support Ed25519 with signatures defined in [RFC8032](https://www.rfc-editor.org/rfc/rfc8032#section-5.1).
+
+Implementations MAY support RSA, Secp256k1 and ECDSA for private use, but peers
+from the public IPFS swarm and DHT may not be able to resolve IPNS records
```

Comment:
Seems mostly reasonable as this copies from the libp2p peerID spec, however RSA keys might still be very common since they used to be the default in implementations like kubo (formerly go-ipfs) and js-ipfs. This is the same as in the libp2p world, but given we're not just referencing that spec directly figured I'd call it out here in case anyone comes looking.

It'd be nice if we could post some numbers here (on GitHub) coming from the Hydra nodes on the distribution of key types justifying that Ed25519 covers most usage.

Reply:
Will try to get %  usage of Ed25519, RSA, Secp256k1 and ECDSA and get back to this with numbers.

Reply:
Some quick numbers from hydras (records that are currently valid)

```
Total DynamoDB Records: 256697
  ipns: 12287
  pk:   244410

DynamoDB Record Value Sizes:
  avg:    299.62269913555673
  std:    45.88040769545226
  median: 299
  max:    2033
  min:    36
IPNS Records Signatures:
  v1:     12287
  v2:     6892
  with pk: 913
PK Records Key Types:
  rsa:       244403
  ed25519:   3
  secp256k1: 0
  ecdsa:     4
```
ed25519 is used by default and inlined, "3" is probably someone writing own  IPNS tool and embedding it the same way RSA - can be ignored.

“with pk” means IPNS records with IpnsEntry.pubKey present.
RSA  is dominating here, and used in ~7% of cases. 

I think Ed25519 being ~93%  and RSA ~7% is a good argument to keep Ed25519 as MUST and change RSA to SHOULD (same as in PeerID spec). 

I adjusted language in https://github.com/ipfs/specs/pull/319/commits/41612da4bebec847612e3efcce83094205903f63